src/ess/reduce/streaming.py

```diff
@@ -280,16 +299,32 @@ def __init__(
             workflow[key] = base_workflow[key]
         for key in dynamic_keys:
             workflow[key] = None  # hack to prune branches
+        for key in context_keys:
+            workflow[key] = None
 
         self._dynamic_keys = set(dynamic_keys)
+        self._context_keys = set(context_keys)
 
         # Find and pre-compute static nodes as far down the graph as possible
-        # See also https://github.com/scipp/sciline/issues/148.
-        nodes = _find_descendants(workflow, dynamic_keys)
-        parents = _find_parents(workflow, nodes) - nodes
-        for key, value in base_workflow.compute(parents).items():
+        nodes = _find_descendants(workflow, dynamic_keys + context_keys)
+        last_static = _find_parents(workflow, nodes) - nodes
+        for key, value in base_workflow.compute(last_static).items():
             workflow[key] = value
 
+        # Nodes that may need updating on context change but should be cached otherwise.
+        dynamic_nodes = _find_descendants(workflow, dynamic_keys)
+        parents_of_dynamic = _find_parents(workflow, dynamic_nodes) - dynamic_nodes
+        context_dependent_parents = parents_of_dynamic & _find_descendants(
+            workflow, context_keys
+        )
+        graph = workflow.underlying_graph
+        self._context_dependencies = {
+            context_key: nx.descendants(graph, context_key) & context_dependent_parents
+            for context_key in self._context_keys
+            if context_key in graph
+        }
+
+        self._context_workflow = workflow.copy()
```

Comment:
This is already an issue with dynamic keys, but is there a way to avoid working on the workflow and instead working with the task graph? This would avoid having to rebuild the task graph on every update.

Reply:
Hmm, great question. It would require setting values on task graphs. Should be possible in principle, and should be considered if performance benchmarks indicate we should.

Reply:
Made a start [here](https://github.com/scipp/essreduce/tree/streaming-no-graph-rebuild), but it was getting slightly too complicated. Waiting for data to decide if it is worth pursuing.